aether-ui/aether/ui/api/serializers.py

```diff
@@ -49,12 +55,47 @@ class PipelineSerializer(DynamicFieldsMixin, serializers.ModelSerializer):
 
     contracts = ContractSerializer(many=True, read_only=True)
 
+    project = serializers.PrimaryKeyRelatedField(
+        required=False,
+        queryset=models.Project.objects.all(),
+    )
+
+    def create(self, validated_data):
+        if not validated_data.get('project'):
+            # assign new pipelines to the default project
+            default_project = models.Project.objects.filter(is_default=True).first()
+            validated_data['project'] = default_project
+
+        instance = super(PipelineSerializer, self).create(validated_data)
+
+        # create initial contract
+        models.Contract.objects.create(
+            pk=instance.pk,      # re-use same uuid
+            name=instance.name,  # re-use same name
+            pipeline=instance,
+        )
+
+        return instance
+
     def update(self, instance, validated_data):
         if instance.contracts.filter(is_read_only=True).count() > 0:
-            raise serializers.ValidationError({'description': _('Input is readonly')})
+            raise serializers.ValidationError({'detail': _('Pipeline is read only')})
```

Comment:
Is this correct? Can a whole Pipeline be read-only? I thought that Inputs and Contracts could be read-only, but that it is always possible to add a new Contract to an existing Pipeline, even if the Input is read-only?

Reply:
You can always add a new contract, what you cannot do is to edit the pipeline, I mean the input JSON data and the AVRO schema.